From f4648065447b628fce3d646cc6627369d36818b4 Mon Sep 17 00:00:00 2001
From: Dylan Yip <dylan.yip@xilinx.com>
Date: Tue, 9 Mar 2021 12:36:50 -0800
Subject: [PATCH 1777/1851] drm: xlnx: zynqmp_dpsub: Only init drm master when
 DP CRTC used

commit 9be4a15c645e00a7377c4a9922cd0ddce13e8882 from
https://github.com/Xilinx/linux-xlnx.git

If an external DRM CRTC is connected to the PS DP subsystem, there will
2 logical drm masters. To avoid this, only initialize drm pipeline if
there is no external CRTC attached.

Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Signed-off-by: Jianqiang Chen <jianqiang.chen@xilinx.com>
Acked-by: Varunkumar Allagadapa <varunkumar.allagadapa@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_dpsub.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_dpsub.c b/drivers/gpu/drm/xlnx/zynqmp_dpsub.c
index 369d111b093c..7197547c9eeb 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dpsub.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_dpsub.c
@@ -97,10 +97,12 @@ static int zynqmp_dpsub_probe(struct platform_device *pdev)
 		goto err_rmem;
 	}
 
-	dpsub->master = xlnx_drm_pipeline_init(pdev);
-	if (IS_ERR(dpsub->master)) {
-		dev_err(&pdev->dev, "failed to initialize the drm pipeline\n");
-		goto err_populate;
+	if (!dpsub->external_crtc_attached) {
+		dpsub->master = xlnx_drm_pipeline_init(pdev);
+		if (IS_ERR(dpsub->master)) {
+			dev_err(&pdev->dev, "failed to initialize the drm pipeline\n");
+			goto err_populate;
+		}
 	}
 
 	dev_info(&pdev->dev, "ZynqMP DisplayPort Subsystem driver probed");
@@ -126,7 +128,8 @@ static int zynqmp_dpsub_remove(struct platform_device *pdev)
 	struct zynqmp_dpsub *dpsub = platform_get_drvdata(pdev);
 	int err, ret = 0;
 
-	xlnx_drm_pipeline_exit(dpsub->master);
+	if (!dpsub->external_crtc_attached)
+		xlnx_drm_pipeline_exit(dpsub->master);
 	of_platform_depopulate(&pdev->dev);
 	of_reserved_mem_device_release(&pdev->dev);
 	component_del(&pdev->dev, &zynqmp_dpsub_component_ops);
-- 
2.31.1

